misc/ls-extensions/modules/performance-analyzer-services/src/main/java/io/ballerina/PerformanceAnalyzerNodeVisitor.java

```diff
@@ -0,0 +1,536 @@
+    private Document document;
+    private boolean withinRange = false;
+
+    public PerformanceAnalyzerNodeVisitor(SemanticModel model, String file, Range range) {
+
+        this.variableMap = new HashMap<>();
+        this.referenceMap = new HashMap<>();
+        this.endPointDeclarationMap = new HashMap<>();
+        this.startNode = new Node();
+        this.currentNode = this.startNode;
+        this.model = model;
+        this.file = file;
+        this.range = range;
+    }
+
+    public void setDocument(Document document) {
+
+        this.document = document;
+    }
+
+    @Override
+    public void visit(ModulePartNode modulePartNode) {
+
+        modulePartNode.members().forEach(moduleMemberDeclarationNode -> moduleMemberDeclarationNode.accept(this));
+    }
+
+    @Override
+    public void visit(ModuleVariableDeclarationNode moduleVariableDeclarationNode) {
+
+        Optional<Symbol> symbol = model.symbol(moduleVariableDeclarationNode);
+        if (symbol.isEmpty() || moduleVariableDeclarationNode.initializer().isEmpty()) {
+            return;
+        }
+
+        boolean clientObject = isClientObject(symbol.get());
+        if (clientObject) {
+            resolveEndPoint(symbol.get(), moduleVariableDeclarationNode.initializer().get());
+        }
+
+        LineRange lineRange = moduleVariableDeclarationNode.typedBindingPattern().bindingPattern().
+                location().lineRange();
+
+        if (clientObject || isRecordObject(symbol.get())) {
+            Optional<ExpressionNode> expressionNode = moduleVariableDeclarationNode.initializer();
+
+            if (expressionNode.isPresent() && expressionNode.get().kind() != SyntaxKind.SIMPLE_NAME_REFERENCE) {
+                this.registerVariableRef(lineRange, moduleVariableDeclarationNode);
+            }
+        } else if ((symbol.get().kind() == SymbolKind.VARIABLE) &&
+                (moduleVariableDeclarationNode.initializer().get().kind() == SyntaxKind.REMOTE_METHOD_CALL_ACTION)) {
+
+            this.registerVariableRef(lineRange, moduleVariableDeclarationNode);
+        }
+
+        if (moduleVariableDeclarationNode.typedBindingPattern().bindingPattern().kind() ==
+                SyntaxKind.CAPTURE_BINDING_PATTERN) {
+
+            SyntaxKind initializerKind = moduleVariableDeclarationNode.initializer().get().kind();
+            if (initializerKind == SyntaxKind.NUMERIC_LITERAL || initializerKind == SyntaxKind.STRING_LITERAL) {
+                this.registerVariableRef(lineRange, moduleVariableDeclarationNode);
+            }
+        }
+        moduleVariableDeclarationNode.typedBindingPattern().accept(this);
+        moduleVariableDeclarationNode.initializer().ifPresent(initializer -> initializer.accept(this));
+    }
+
+    @Override
+    public void visit(FunctionDefinitionNode functionDefinitionNode) {
+
+        LineRange lineRange = functionDefinitionNode.lineRange();
+        if (functionDefinitionNode.syntaxTree().filePath().equals(file)
```

Comment:
Why do we access the syntax tree here? Can't we ensure that we are traversing the particular tree by extracting the tree first?